manpages/mrcbyte.man
.na
.nh
.TH mrcbyte 1 2.42 BL3DEMC
.SH NAME
mrcbyte \- Converts an MRC file into byte type data.
.SH SYNOPSIS
mrcbyte  [options]  input_file  output_file
.SH DESCRIPTION
mrcbyte will convert MRC files to an MRC byte type file.  The
default is to convert the entire file with a full linear ramp
and write the output to byte type data.  With appropriate options,
the intensity scaling can be modified to match the scaling achieved when 
viewing the file in 3dmod(1).  Subsets of the image can also be copied.
Complex data will also be scaled just as when loaded into 3dmod, so this
program can be used to save a viewable form of a Fourier transform as bytes.
.SH OPTIONS
.TP
.B -c \fIblack,white\fR
Specify intensity scaling to match the scaling achieved with the given
black and white contrast settings in 3dmod(1).
.TP
.B -s \fImin,max\fR
Specify an initial intensity scaling, which will scale "min" to 0 and
"max" to 255, just as is achieved with the 
.B -s
option in 3dmod.  If both
.B -c
and
.B -s
are entered, the final scaling will be the product of these two scalings.
.TP
.B -l
Use logarithmic scaling (valid for short integer and floating point data).
.TP
.B -e
Use exponential scaling (valid for short integer and floating point data).
.TP
.B -R
Reverse contrast in the output.
.TP
.B -x \fIX0,X1\fR
Write out pixels in X between X0 and X1, inclusive, numbered from 0.
.TP
.B -y \fIY0,Y1\fR
Write out pixels in Y between Y0 and Y1, inclusive, numbered from 0.
.TP
.B -z \fIZ0,Z1\fR
Write out pixels in Z between Z0 and Z1, inclusive, numbered from 0.
.TP
.B -b
Write raw byte data without a header.
.TP
.B -r
Request user input for the contrast and size options.
.SH EXAMPLES
.nf
Example of a manual mrc resizing operation.

% mrcbyte -r image.mrc byte.mrc 
Enter (min x, max x). (return for default) >10 100
Enter (min y, max y). (return for default)  >10 100
Enter sections (low, high)  >0 1
Enter (black level, white level) >34 200

Converting Image # 002

Example of a command that does the same thing.

% mrcbyte -x 10,100 -y 10,100 -z 0,1 -c 34,200 image.mrc byte.mrc
.fi	
.SH BUGS
There is no protection against taking the log of numbers <= 0 or taking
exponentials of numbers that are too large.

.SH AUTHORS
Jim Kremer and David Mastronarde  <mast@colorado.edu>
.SH SEE ALSO
3dmod(1), trimvol(1), newstack(1)


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)


mrc/mrcbyte.c
/*
 *  mrcbyte.c -- Converts an mrc file to byte format.
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include "mrcfiles.h"
#include "mrcslice.h"
#include "b3dutil.h"

void mrcbyte_help(char *name)
{
  printf("Usage: %s -[rRblecsxyz] <infile> <outfile>\n", name);
  printf("Options:\n");
  printf("\t-r\tRequest user input for size and contrast info.\n");
  printf("\t-c blk,wht\tContrast scaling with black and white levels\n");
  printf("\t-s min,min\tInitial intensity scaling (min to 0, max to 255)\n");
  printf("\t-R\tReverse contrast in output.\n");
  printf("\t-l\tUse logarithmic ramp.\n");
  printf("\t-e\tUse exponential ramp.\n");
  printf("\t-x min,max\tWrite subset of image in X\n");
  printf("\t-y min,max\tWrite subset of image in Y\n");
  printf("\t-z min,max\tWrite subset of image in Z\n");
  printf("\t-b\tWrite raw byte data, no header\n");
}

int main( int argc, char *argv[] )
{

  int    i = 0, k;
  FILE   *fin, *fout;
  int resize = FALSE;
  int ramptype = MRC_RAMP_LIN;
  MrcHeader hdata, hout;
  IloadInfo  li;
  Islice slice;
  char line[128];
  int mode = 0;
  int data_only = FALSE;
  int reverse_video = FALSE;
  float xscl, yscl, zscl;
  size_t xysize, ii;
  unsigned char *buf;
  double meansum = 0.;
  char *progname = imodProgName(argv[0]);


  if (argc < 2){
    printf("%s version %s\n", progname, VERSION_NAME);
    imodCopyright();
    mrcbyte_help(progname);
    exit(3);
  }

  /* Make library error output to stderr go to stdout */
  b3dSetStoreError(-1);

  mrc_init_li(&li, NULL);

  for (i = 1; i < argc; i++)
    if (argv[i][0] == '-')
      switch (argv[i][1]){
          
      case 'b':
        data_only = TRUE;
        break;

      case 'r':
        resize = TRUE;
        break;

      case 'R':
        reverse_video = TRUE;
        break;

      case 'l':
        ramptype = MRC_RAMP_LOG;
        break;

      case 'e':
        ramptype = MRC_RAMP_EXP;
        break;

      case 'c':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-c%d%*c%d", &(li.black), &(li.white));
        else
          sscanf(argv[++i], "%d%*c%d", &(li.black), &(li.white));
        break;

        /* DNM 2/15/01: add scaling option to replicate imod, but
           eliminate short and float options */
      case 's':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-s%f%*c%f", &(li.smin), &(li.smax));
        else
          sscanf(argv[++i], "%f%*c%f", &(li.smin), &(li.smax));
        break;

      case 'x':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-x%d%*c%d", &(li.xmin), &(li.xmax));
        else
          sscanf(argv[++i], "%d%*c%d", &(li.xmin), &(li.xmax));
        break;
            
      case 'y':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-y%d%*c%d", &(li.ymin), &(li.ymax));
        else
          sscanf(argv[++i], "%d%*c%d", &(li.ymin), &(li.ymax));
        break;
            
      case 'z':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-z%d%*c%d", &(li.zmin), &(li.zmax));
        else
          sscanf(argv[++i], "%d%*c%d", &(li.zmin), &(li.zmax));
        break;

      default:
        printf("ERROR: %s - Illegal option %s\n", progname, argv[i]);
        mrcbyte_help(progname);
        exit(1);
        break;
            
      }else break;

  if (i < (argc - 1)){
      

    fin = fopen(argv[i], "rb");
    if (fin == NULL) {
      printf("ERROR: %s - Opening %s.\n", progname, argv[i]);
      exit(3);
    }
 
    i++;
    if (!getenv("IMOD_NO_IMAGE_BACKUP") && imodBackupFile(argv[i])) {
      printf("WARNING: %s - Error making backup file from existing %s\n",
             progname, argv[i]);
    }
    fout = fopen(argv[i], "wb");
    if (fout == NULL) {
      printf("ERROR: %s - Opening %s.\n", progname, argv[i]);
      exit(3);
    }
  }else{
    mrcbyte_help(progname);
    exit(3);     
  }
     

  if (mrc_head_read(fin, &hdata)) {
    printf("ERROR: %s - Reading input file header.\n", progname);
    exit(3);
  }
     
  if (resize){
    printf("Defaults for %s size = ( %d x %d x %d):\n", 
           argv[i - 1], hdata.nx, hdata.ny, hdata.nz);
    printf("x = ( 0, %d), y = ( 0, %d), z = ( 0, %d), c = ( 0, 255)\n",
           hdata.nx - 1, hdata.ny - 1, hdata.nz - 1);
    resize = get_loadinfo(&hdata, &li);
    printf (" Enter (black level, white level) >");
    fgetline(stdin, line, 127);
    sscanf(line, "%d%*c%d\n", &(li.black), &(li.white));
    printf("\n");
    if ( (!li.black) & (!li.white))
      li.white = 255;     
  }

  li.ramp = ramptype;
  if (ramptype != MRC_RAMP_LIN && hdata.mode != MRC_MODE_SHORT &&
      hdata.mode != MRC_MODE_USHORT && hdata.mode != MRC_MODE_FLOAT) {
    printf("ERROR: %s - Nonlinear scaling can be used only with integer and "
           "float data\n", progname);
    exit(3);
  }

  /* This takes care of fixing the mins and maxes */
  mrc_init_li(&li, &hdata);

  /* This sets the scaling */
  mrcContrastScaling(&hdata, li.smin, li.smax, li.black, li.white, li.ramp, 
                     &li.slope, &li.offset);
  
  hout = hdata;
  hout.amin = 0.;
  hout.amax = 255.;
  hout.mode = mode;

  /* DNM: eliminate extra header info in the output, and mark it as not
     swapped now that we're done reading data */
  mrcInitOutputHeader(&hout);
  hout.nx = li.xmax + 1 - li.xmin;
  hout.ny = li.ymax + 1 - li.ymin;
  hout.nz = li.zmax + 1 - li.zmin;
  mrc_get_scale(&hdata, &xscl, &yscl, &zscl);
  mrc_set_scale(&hout, (double)xscl, (double)yscl, (double)zscl);

  mrc_head_label(&hout, "mrcbyte: Converted and scaled to byte mode.");

  xysize = (size_t)hout.nx * (size_t)hout.ny;
  buf = (unsigned char *)malloc(xysize);
  if (!buf) {
    printf("ERROR: %s - Allocating memory to read data into\n", progname);
    exit(1);
  }

  if (!data_only && mrc_head_write(fout, &hout)) {
    printf("ERROR: %s - Writing header to output file\n", progname);
    exit(1);
  }

  for (k = 0; k < hout.nz; k++) {
    printf("Converting Image # %3d\r", k + li.zmin);
    fflush(stdout);
    if (mrcReadZByte(&hdata, &li, buf, k + li.zmin)) {
      printf("ERROR: %s - Reading section %d from file\n", progname,k+li.zmin);
      exit(1);
    }
    if (reverse_video){
      for (ii = 0; ii < xysize; ii++)
        buf[ii] = (unsigned char)(255 - (int)buf[ii]);
    }
    
    sliceInit(&slice, hout.nx, hout.ny, SLICE_MODE_BYTE, buf);
    sliceMMM(&slice);
    if (!k) {
      hout.amin = slice.min;
      hout.amax = slice.max;
    } else {
      hout.amin = B3DMIN(hout.amin, slice.min);
      hout.amax = B3DMAX(hout.amax, slice.max);
    }
    meansum += slice.mean;
    if (mrc_write_slice(buf, fout, &hout, k, 'Z')) {
      printf("ERROR: %s - Writing section %d to file\n", progname, k);
      exit(1);
    }

      
  }
  hout.amean = meansum / hout.nz;

  if (!data_only && mrc_head_write(fout, &hout)) {
    printf("ERROR: %s - Writing header to output file\n", progname);
    exit(1);
  }
  fclose(fout);
  printf("\nDone\n");

  exit(0);
}
